### PR TITLE
mgr/volumes: remove incorrect 'size' from output of 'snapshot info'

### PR DESCRIPTION
**Pull request by nmshelke**
The 'size' shown in the output of snapshot info command relies on
rstats which is incorrect snapshot size. It tracks size of the
subvolume from the snapshot has been taken instead of the snapshot
itself. Hence having the 'size' field in the output of 'snapshot info'
doesn't make sense until the rstats is fixed.

Fixes: https://tracker.ceph.com/issues/55822
Signed-off-by: Nikhilkumar Shelke <nshelke@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by nmshelke** [comment]
@vshankar Please find teuthology link with this changes: https://pulpito.ceph.com/nshelke-2022-06-12_14:49:47-fs:volumes-wip-nshelke-remove-size-from-snapshot-info-distro-default-smithi/

**Review by anthonyeleven — APPROVED** [review]
Docs LGTM, the code changes are beyond my ken, so a dev should approve those before merging.

**Review by vshankar — APPROVED** [review]

**Comment by Madhu-1** [comment]
> > @nmshelke Did you double check about the usage of snapshot size with the CSI team ?
> 
> @Madhu-1 Could you please confirm if we remove `size` from `fs subvolume snapshot info` command then it will not break any existing functionality?

We are not using it in cephcsi **but** we need the size field in the snapshot details. we already have a bug if we consider subvolume size as the snapshot size in cephcsi https://github.com/ceph/ceph-csi/issues/3147. When cephcsi creates the snapshot, it needs to return the size of the snapshot. The exact size will be used to validate during clone operation as we don't allow users to create smaller clones as it can cause data loss.

**Comment by kotreshhr** [comment]
> > > @nmshelke Did you double check about the usage of snapshot size with the CSI team ?
> > 
> > 
> > @Madhu-1 Could you please confirm if we remove `size` from `fs subvolume snapshot info` command then it will not break any existing functionality?
> 
> We are not using it in cephcsi **but** we need the size field in the snapshot details. we already have a bug if we consider subvolume size as the snapshot size in cephcsi [ceph/ceph-csi#3147](https://github.com/ceph/ceph-csi/issues/3147). When cephcsi creates the snapshot, it needs to return the size of the snapshot. The exact size will be used to validate during clone operation as we don't allow users to create smaller clones as it can cause data loss.

@Madhu-1 You mean to say, the CSI is planning to use the `size` field from the output of `snapshot info` in the future but not using it now ? But the snapshot size is incorrect and there is no easy way to fetch it unless the rstats is fixed which isn't trivial.

**Comment by Madhu-1** [comment]
> > > > @nmshelke Did you double check about the usage of snapshot size with the CSI team ?
> > > 
> > > 
> > > @Madhu-1 Could you please confirm if we remove `size` from `fs subvolume snapshot info` command then it will not break any existing functionality?
> > 
> > 
> > We are not using it in cephcsi **but** we need the size field in the snapshot details. we already have a bug if we consider subvolume size as the snapshot size in cephcsi [ceph/ceph-csi#3147](https://github.com/ceph/ceph-csi/issues/3147). When cephcsi creates the snapshot, it needs to return the size of the snapshot. The exact size will be used to validate during clone operation as we don't allow users to create smaller clones as it can cause data loss.
> 
> @Madhu-1 You mean to say, the CSI is planning to use the `size` field from the output of `snapshot info` in the future but not using it now ? But the snapshot size is incorrect and there is no easy way to fetch it unless the rstats is fixed which isn't trivial.

Yes planning to use `size` from the `snapshot info` command in CephCSI. In Ceph FS No way to store the snapshot size as metadata once the snapshot is created?

**Comment by kotreshhr** [comment]
> > > > > @nmshelke Did you double check about the usage of snapshot size with the CSI team ?
> > > > 
> > > > 
> > > > @Madhu-1 Could you please confirm if we remove `size` from `fs subvolume snapshot info` command then it will not break any existing functionality?
> > > 
> > > 
> > > We are not using it in cephcsi **but** we need the size field in the snapshot details. we already have a bug if we consider subvolume size as the snapshot size in cephcsi [ceph/ceph-csi#3147](https://github.com/ceph/ceph-csi/issues/3147). When cephcsi creates the snapshot, it needs to return the size of the snapshot. The exact size will be used to validate during clone operation as we don't allow users to create smaller clones as it can cause data loss.
> > 
> > 
> > @Madhu-1 You mean to say, the CSI is planning to use the `size` field from the output of `snapshot info` in the future but not using it now ? But the snapshot size is incorrect and there is no easy way to fetch it unless the rstats is fixed which isn't trivial.
> 
> Yes planning to use `size` from the `snapshot info` command in CephCSI. In Ceph FS No way to store the snapshot size as metadata once the snapshot is created?

Ok, then we need to think how to get around this. One way is to store the subvolume size at the time of snapshot creation  in the metadata file. 

@vshankar what do you think ?

**Comment by vshankar** [comment]
> Ok, then we need to think how to get around this. One way is to store the subvolume size at the time of snapshot creation in the metadata file.
> 
> @vshankar what do you think ?

If we can guarantee that the clients flush outstanding data on snapshot creation, then, this sounds fine. And the size will be fetched via rstats anyway, isn't it? (though, it will be for the HEAD version).

**Comment by Madhu-1** [comment]
will discuss this in a cephcsi call tomorrow. and update this thread.

**Comment by vshankar** [comment]
> > > Ok, then we need to think how to get around this. One way is to store the subvolume size at the time of snapshot creation in the metadata file.
> > > @vshankar what do you think ?
> > 
> > 
> > If we can guarantee that the clients flush outstanding data on snapshot creation, then, this sounds fine. And the size will be fetched via rstats anyway, isn't it? (though, it will be for the HEAD version).
> 
> Currently, there is no guarantee for it :( we have support for only crash-consistent snapshots, not for the app-consistent snapshots in CSI. but in the future, there is a plan to add webhooks to pods to flush data before taking snapshots,s, etc.

Question is - is it fine for csi is size is not accurate?

**Comment by Madhu-1** [comment]
> > > > Ok, then we need to think how to get around this. One way is to store the subvolume size at the time of snapshot creation in the metadata file.
> > > > @vshankar what do you think ?
> > > 
> > > 
> > > If we can guarantee that the clients flush outstanding data on snapshot creation, then, this sounds fine. And the size will be fetched via rstats anyway, isn't it? (though, it will be for the HEAD version).
> > 
> > 
> > Currently, there is no guarantee for it :( we have support for only crash-consistent snapshots, not for the app-consistent snapshots in CSI. but in the future, there is a plan to add webhooks to pods to flush data before taking snapshots,s, etc.
> 
> Question is - is it fine for csi is size is not accurate?

as it's not accurate it's not useful for CSI. currently, we are using subvolume size as snapshot-size we will use the same until the size issue is fixed for a snapshot.

**Comment by Madhu-1** [comment]
> will discuss this in a cephcsi call tomorrow. and update this thread.

We discussed in csi call today, that we can remove the size field.

**Comment by nmshelke** [comment]
> > will discuss this in a cephcsi call tomorrow. and update this thread.
> 
> We discussed in csi call today, that we can remove the size field.

Thanks @Madhu-1 for updates.

**Review by ajarr — APPROVED** [review]

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-June-20

**Pull request merged**
